File: backend/src/main/java/com/runningcrew/common/web/PageResponse.java
package com.runningcrew.common.web;

import java.util.List;
import org.springframework.data.domain.Page;

/**
 * 목록 응답 공통 래퍼(계약 conventions.md §6, offset 기반).
 *
 * <p>JSON 필드는 snake_case: {@code items, page, size, total_elements, total_pages}.
 *
 * @param <T> 요소 타입(각 목록 계약에서 정의)
 */
public record PageResponse<T>(
        List<T> items,
        int page,
        int size,
        long totalElements,
        int totalPages
) {
    /** Spring Data {@link Page}를 계약 래퍼로 변환한다. */
    public static <T> PageResponse<T> from(Page<T> page) {
        return new PageResponse<>(
                page.getContent(),
                page.getNumber(),
                page.getSize(),
                page.getTotalElements(),
                page.getTotalPages());
    }
}

File: app/lib/core/model/page_response.dart
/// 페이지네이션 공통 래퍼 (conventions.md §6, offset 방식 확정 v0.1.1).
class PageResponse<T> {
  const PageResponse({
    required this.items,
    required this.page,
    required this.size,
    required this.totalElements,
    required this.totalPages,
  });

  final List<T> items;
  final int page; // 0-base
  final int size;
  final int totalElements;
  final int totalPages;

  factory PageResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic>) itemFromJson,
  ) =>
      PageResponse(
        items: [
          for (final item in (json['items'] as List<dynamic>? ?? const []))
            itemFromJson(item as Map<String, dynamic>),
        ],
        page: json['page'] as int? ?? 0,
        size: json['size'] as int? ?? 0,
        totalElements: json['total_elements'] as int? ?? 0,
        totalPages: json['total_pages'] as int? ?? 0,
      );
}
